### PR TITLE
add dedicated structures for PKCS12 and JKS stores

### LINKED ISSUE #199: Support of setting arbitrary password for PKCS12 truststore
We are using cert-manager to issue certificates to various-servers running in our Kubernetes cluster. Currently we use  truststore.p12 in the placed in kubernetes Secret to establish trust to the client java applications (sTLS). 
Our java applications are running on java11 and [java17](https://bugs.openjdk.org/browse/JDK-8275252) which does not supports password-less pkcs12 container as JVM argument.

To fix this, we were hoping to use this cert-manager component to distribute the trust. But since a all of our clients are Java-based, we need to be able to provide the trust bundle in password protected PKCS12 format. 

_-Djavax.net.ssl.trustStore=/configuration/p12/truststore/trustStore.p12 -Djavax.net.ssl.trustStorePassword=xxxxxx_

Would it be possible to add an opt-in to provide the arbitrary password to protected PKCS12?

### PR DESCRIPTION
**Pull request by arsenalzp**
This PR fixes #199  .
As was proposed by @erikgb, two dedicated structures PKCS12 and JKS were added.
Could you please be so kind to review this PR?

### CODE REVIEW

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:140]:
```suggestion
        // +optional
	//+kubebuilder:validation:MaxLength=30
	//+kubebuilder:default=""
	Password *string `json:"password"`
```

**Review by erikgb — CHANGES_REQUESTED** [review]
@arsenalzp Please try to avoid closing PRs and opening new ones. It makes it harder to track work on this new feature.

I still think the API additions need a bit more work, and again I think this is the most important part. Making a field with a default required does not make sense to me. Though a password IS required, it has a default. So it is not required from the user's point of view. I've put up some suggestions, but I am not sure about the min/max length requirements. It is not easy to find any good specification of these legacy formats.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (arsenalzp) [reply pkg/apis/trust/v1alpha1/types_bundle.go:134]: Fixed.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (arsenalzp) [reply pkg/apis/trust/v1alpha1/types_bundle.go]: Fixed

**Comment by arsenalzp** [comment]
Hello colleagues,
Should I put an improvement in current PR? If so, could you be so kind to put remarks 🙂

**Comment by erikgb** [comment]
> Hello colleagues, Should I put an improvement in current PR? If so, could you be so kind put your remarks 🙂

Please rebase the PR and ensure you have run `make generate`. I still see the password field marked as required in the generated OpenAPI schema, and that's not correct. And another tip, at least when I am the reviewer: please resolve conversations when they are fixed! If needed they can be reopened. It just makes the review process more comfortable for all parties.

File: pkg/bundle/bundle_test.go
Comment (erikgb) [inline pkg/bundle/bundle_test.go:571]:
```suggestion
							Password: ptr.To(DefaultJKSPassword),
```

**Review by erikgb — CHANGES_REQUESTED** [review]
Added a few suggestions. I wonder if we should **not** use the default passwords in tests. Ideally, the defaults should be just be present in the API spec.

File: pkg/bundle/sync.go
Comment (arsenalzp) [reply pkg/bundle/sync.go:85]: I guess to have structures, global vars and consts as well as imports in one place rather than somewhere in the middle of code lines.
If you think that was bad decision then I can move it back.

Comment (erikgb) [reply pkg/bundle/sync.go:85]: I think we should avoid moving it in this PR. I personally prefer to have the structs defined together with the methods having the structs as receiver. But the main point here is to not move things just out of personal preference.

**Comment by arsenalzp** [comment]
All remarks were fixed, branch was re-based as well.

File: pkg/bundle/sync.go
Comment (erikgb) [inline pkg/bundle/sync.go:267]:
```suggestion
	return pkcs12.LegacyRC2.EncodeTrustStoreEntries(entries, e.password)
```

I think we should be conservative about changing the crypto in this PR. Ref. https://github.com/cert-manager/cert-manager/pull/6517

**Review by erikgb — CHANGES_REQUESTED** [review]
Nice, thanks for this. Almost LGTM from me. Are we able to add some tests using this new feature? It seems like all tests use the default passwords.

**Comment by arsenalzp** [comment]
> Nice, thanks for this. Almost LGTM from me. Are we able to add some tests using this new feature? It seems like all tests use the default passwords.

Hello,
Yes, I will add tests for both stores, which use arbitrary password.
Is that correct?

**Comment by arsenalzp** [comment]
I spend the whole evening with troubleshooting PKCS12 test case, and I've just found that PKCS12 encoders are not deterministic:
```
$ openssl pkcs12 -nokeys -in test.pem -export -out test.p12 -password pass:qwerty
$ md5sum test.p12
9176d1b36d911766e74810ff0fa3bbcd  test.p12
$ openssl pkcs12 -nokeys -in test.pem -export -out test.p12 -password pass:qwerty
$ md5sum test.p12
```
89b666d7907c0b8420470f1c25a66c2b  test.p12
[arsenal@arsenal-laptop pkcs12]$

**Comment by erikgb** [comment]
> I spent the whole evening with troubleshooting PKCS12 test case, and I've just found that PKCS12 encoders are not deterministic:

Yes, both JKS and PKCS12 adds some salt making them non-deterministic. I think the tests must be created with that in mind. What's wrong with the existing test approach?

https://github.com/cert-manager/trust-manager/blob/01bd331abb8ee071025e2b8989930a2eb3b1d8e9/pkg/bundle/sync_test.go#L1541-L1573

**Comment by arsenalzp** [comment]
> > I spent the whole evening with troubleshooting PKCS12 test case, and I've just found that PKCS12 encoders are not deterministic:
> 
> Yes, both JKS and PKCS12 adds some salt making them non-deterministic. I think the tests must be created with that in mind. What's wrong with the existing test approach?
> 
> https://github.com/cert-manager/trust-manager/blob/01bd331abb8ee071025e2b8989930a2eb3b1d8e9/pkg/bundle/sync_test.go#L1541-L1573

Thank you for a hint!
I was trying to put test in bundle_test.go. 
Now it works!

File: pkg/bundle/sync_test.go
Comment (erikgb) [inline pkg/bundle/sync_test.go:1333]:
```suggestion
		expPassword    *string
```

I wonder if this could simplify the test? If JKS/PKCS12 expected: if `expPassword` is set, use the value, otherwise use default password? That would remove the need for a new constant.

**Review by erikgb — APPROVED** [review]
Thanks @arsenalzp! Did a quick review now. Just one suggestion. I would like to take a closer look and another review from a maintainer before we merge. 

Fixes #199 

/approve
/hold

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:141]:
```suggestion
	// Password to PKCS12 trust store
```

**Review by inteon — COMMENTED** [review]
LGTM, just one small typo

**Comment by arsenalzp** [comment]
> @arsenalzp Maybe you can update the PR title and description? It seems like I am not allowed to do it. I suggest;
> 
>     * rewording the title to describe the feature added
> 
>     * change "relates to" to "fixes" [Support of setting arbitrary password for PKCS12 truststore #199](https://github.com/cert-manager/trust-manager/issues/199) (so the issue will be closed on merge of this PR)

Done!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:120]: suggestion: Looks like this comment will need updating. Might be worth linking to https://cert-manager.io/docs/faq/#keystore-passwords as well!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:123]: suggestion: Looks like this comment will need updating. Might be worth linking to https://cert-manager.io/docs/faq/#keystore-passwords as well!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:133]: suggestion: It's good to have _some_ max length, but I think 30 might be a bit small. A reasonable password to use here would be 32 hex encoded bytes (`openssl rand -hex 32`), which would have a length of 64 bytes. I'd suggest setting the max length to 128 so that there is a max but if someone wants to use a really secure password they can.

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:143]: suggestion: As above, I'd suggest increasing this max length!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:130]: nitpick:

```suggestion
	// Password for JKS trust store
```

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (SgtCoDFish) [inline pkg/apis/trust/v1alpha1/types_bundle.go:141]: nitpick:

```suggestion
	// Password for PKCS12 trust store
```

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (arsenalzp) [reply pkg/apis/trust/v1alpha1/types_bundle.go:130]: Ooopps, my mistake, will fix!

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (arsenalzp) [reply pkg/apis/trust/v1alpha1/types_bundle.go:133]: Fixed.

**Review by inteon — COMMENTED** [review]
/approve
/lgtm
/hold in case someone else wants to review

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:120]:
```suggestion
	// The bundle has "changeit" as the default password.
```

File: pkg/apis/trust/v1alpha1/types_bundle.go
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:124]:
```suggestion
	// The bundle is by default created without a password.
```

**Review by erikgb — COMMENTED** [review]
A couple of suggestions to API docs improvements, but

/lgtm

**Comment by arsenalzp** [comment]
> A couple of suggestions to API docs improvements, but
> 
> /lgtm

Fixes were implemented.

File: pkg/bundle/sync_test.go
Comment (erikgb) [inline pkg/bundle/sync_test.go:1626]: If you apply the suggested changes above, I think this password logic could be simplified and added to the existing expJKS/expPKCS12 below. WDYT? I would prefer to not modify the bundle resource in the assert section.

Comment (arsenalzp) [reply pkg/bundle/sync_test.go:1626]: Ok, I agree with your suggestions

Comment (arsenalzp) [reply pkg/bundle/sync_test.go:1624]: Suggestions were implemented.

File: pkg/bundle/sync_test.go
Comment (arsenalzp) [reply pkg/bundle/sync_test.go]: Suggestions were implemented.

File: pkg/bundle/sync_test.go
Comment (arsenalzp) [reply pkg/bundle/sync_test.go]: Suggestions were implemented.

**Pull request merged**
